src/backend/distributed/executor/local_executor.c

```diff
@@ -468,49 +468,44 @@ ExtractLocalAndRemoteTasks(bool readOnly, List *taskList, List **localTaskList,
 		/* either the local or the remote should be non-nil */
 		Assert(!(localTaskPlacementList == NIL && remoteTaskPlacementList == NIL));
 
-		if (list_length(task->taskPlacementList) == 1)
+		if (list_length(localTaskPlacementList) > 0)
 		{
-			/*
-			 * At this point, the task has a single placement (e.g,. anchor shard
-			 * is distributed table's shard). So, it is either added to local or
-			 * remote taskList.
-			 */
-			if (localTaskPlacementList == NIL)
+			if (list_length(task->taskPlacementList) == 1)
 			{
-				*remoteTaskList = lappend(*remoteTaskList, task);
+				*localTaskList = lappend(*localTaskList, task);
 			}
 			else
 			{
-				*localTaskList = lappend(*localTaskList, task);
-			}
-		}
-		else
-		{
-			/*
-			 * At this point, we're dealing with reference tables or intermediate
-			 * results where the task has placements on both local and remote
-			 * nodes. We always prefer to use local placement, and require remote
-			 * placements only for modifications.
-			 */
-			task->partiallyLocalOrRemote = true;
+				/*
+				 * At this point, we're dealing with reference tables or intermediate
+				 * results where the task has placements on both local and remote
+				 * nodes. We always prefer to use local placement, and require remote
+				 * placements only for modifications.
+				 */
+				task->partiallyLocalOrRemote = true;
 
-			Task *localTask = copyObject(task);
+				Task *localTask = copyObject(task);
```

Comment:
Thinking of if it is still possible to have "unrecognized node type" errors as we will still be copying `Task` struct at some point. Can we improve `CopyNodeTask` method for `TaskQueryString` ?

Reply:
Changing "queryStringList" to store `StringInfo`s instead of `char*`s should fix possible further issues like the one explained in #3801.

Reply:
As far as I see StringInfo will have the same problem, looking at postgres code it seems that there is already a copy function for T_String which is of type `Value`. 

We could alternatively have a function that copies a char* list in copyfuncs.

Also I have verified that the following fails:

```sql
set citus.shard_replication_factor to 2;
create table dist (a int);
select create_distributed_table('dist', 'a');
begin;
INSERT INTO dist SELECT a + 1 FROM dist_table;
-- here the copy will fail.
```

Reply:
Also this shouldn't be just queryStringList but also perPlacementQueryStrings. So it might be simpler to introduce a function that copies a string list.

Reply:
> Also this shouldn't be just queryStringList but also perPlacementQueryStrings. So it might be simpler to introduce a function that copies a string list.

I am ok with this to prevent major change